File: app/src/main/java/com/cyl/musiclake/api/baidu/BaiduApiServiceImpl.java
package com.cyl.musiclake.api.baidu;

import com.cyl.musiclake.api.ApiManager;
import com.cyl.musiclake.data.model.Music;
import com.cyl.musiclake.ui.common.Constants;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;

/**
 * Created by yonglong on 2018/1/21.
 */

public class BaiduApiServiceImpl {
//    http://musicapi.qianqian.com/v1/restserver/ting?from=android&version=6.0.7.1&channel=huwei&operator=1&method=baidu.ting.billboard.billCategory&format=json&kflag=2

    public static Observable<BaiduMusicList> getOnlinePlaylist() {
        Map<String, String> params = new HashMap<>();
        params.put(Constants.PARAM_METHOD, Constants.METHOD_CATEGORY);// key
        params.put("operator", "1");
        params.put("kflag", "2");
        params.put("format", "json");
        return ApiManager.getInstance().apiService.getOnlinePlaylist(Constants.BASE_URL_BAIDU_MUSIC, params)
                .flatMap(Observable::fromArray);
    }

    public static Observable<List<Music>> getOnlineSongs(String type, int limit, int mOffset) {
        Map<String, String> params = new HashMap<>();

        params.put(Constants.PARAM_METHOD, Constants.METHOD_GET_MUSIC_LIST);
        params.put(Constants.PARAM_TYPE, type);
        params.put(Constants.PARAM_SIZE, String.valueOf(limit));
        params.put(Constants.PARAM_OFFSET, String.valueOf(mOffset));

        return ApiManager.getInstance().apiService.getOnlineSongs(Constants.BASE_URL_BAIDU_MUSIC, params)
                .flatMap(baiduSongList -> {
                    List<Music> musicList = new ArrayList<>();
                    for (BaiduMusicInfo songInfo : baiduSongList.getSong_list()) {
                        Music music = new Music();
                        music.setType(Music.Type.BAIDU);
                        music.setOnline(true);
                        music.setId(songInfo.getSong_id());
                        music.setAlbum(songInfo.getAlbum_title());
                        music.setAlbumId(songInfo.getAlbum_id());
                        music.setArtist(songInfo.getArtist_name());
                        music.setArtistId(songInfo.getArtist_id());
                        music.setTitle(songInfo.getTitle());
                        music.setLrcPath(songInfo.getLrclink());
                        music.setCoverSmall(songInfo.getPic_small());
                        music.setCoverUri(songInfo.getPic_big());
                        music.setCoverBig(songInfo.getPic_radio());
                        musicList.add(music);
                    }
                    return Observable.fromArray(musicList);
                });
    }

    //{"errorCode":22232,"data":{"xcode":"","songList":""}}
    public static Observable<Music> getTingSongInfo(String mid) {
        Map<String, String> params = new HashMap<>();
        String Url = "http://music.baidu.com/data/music/links?songIds=" + mid;
        return ApiManager.getInstance().apiService.getTingSongInfo(Url, params)
                .flatMap(baiduSongInfo -> {
                    Music music = new Music();
                    BaiduSongInfo.DataBean.SongListBean songInfo = baiduSongInfo.getData().getSongList().get(0);
                    music.setType(Music.Type.BAIDU);
                    music.setId(songInfo.getSongId());
                    music.setAlbum(songInfo.getAlbumName());
                    music.setAlbumId(songInfo.getArtistId());
                    music.setArtistId(songInfo.getArtistId());
                    music.setArtist(songInfo.getArtistName());
                    music.setTitle(songInfo.getSongName());
                    music.setUri(songInfo.getSongLink());
                    music.setFileSize(songInfo.getSize());
                    music.setLrcPath(songInfo.getLrcLink());
                    music.setCoverSmall(songInfo.getSongPicSmall());
                    music.setCoverUri(songInfo.getSongPicBig());
                    music.setCoverBig(songInfo.getSongPicRadio());

                    return Observable.create((ObservableOnSubscribe<Music>) e -> {
                        if (music.getUri() != null) {
                            e.onNext(music);
                            e.onComplete();
                        } else {
                            e.onError(new Throwable());
                        }
                    });
                });
    }

    @SuppressWarnings({"unchecked", "varargs"})
    public static Observable<String> getBaiduLyric(String lyricPath) {
        return ApiManager.getInstance().apiService.getXiamiLyric(lyricPath)
                .flatMap(xiaMiLyricInfo -> {
                    String lyric = xiaMiLyricInfo.string();
                    return Observable.fromArray(lyric);
                });
    }

}

File: app/src/main/java/com/cyl/musiclake/api/baidu/BaiduApiServiceImpl.kt
package com.cyl.musiclake.api.baidu

import com.cyl.musicapi.baidu.BaiduApiService
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.bean.Playlist
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.net.ApiManager
import com.cyl.musiclake.utils.FileUtils
import com.cyl.musiclake.utils.LogUtil
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import java.util.*

/**
 * Created by yonglong on 2018/1/21.
 */

object BaiduApiServiceImpl {
    private val TAG = "BaiduApiServiceImpl"
    private val apiService: BaiduApiService
        get() = ApiManager.getInstance().create(BaiduApiService::class.java, Constants.BASE_URL_BAIDU_MUSIC)

    //    http://musicapi.qianqian.com/v1/restserver/ting?from=android&version=6.0.7.1&channel=huwei&operator=1&method=baidu.ting.billboard.billCategory&format=json&kflag=2

    //获取榜单
    fun getOnlinePlaylist(): Observable<List<Playlist>> {
        val params = HashMap<String, String>()
        params[Constants.PARAM_METHOD] = Constants.METHOD_CATEGORY
        params["operator"] = "1"
        params["kflag"] = "2"
        params["format"] = "json"
        return apiService.getOnlinePlaylist(params).flatMap { it ->
            val playlists = mutableListOf<Playlist>()
            for (item in it.content!!) {
                val playlist = Playlist()
                playlist.name = item.name
                playlist.id = item.type.toString()
                playlist.coverUrl = item.picS192
                val musicList = mutableListOf<Music>()
                for (itemMusic in item.content!!) {
                    val music = Music()
                    music.title = itemMusic.title
                    music.album = itemMusic.albumTitle
                    music.artist = itemMusic.author
                    music.albumId = itemMusic.albumId
                    music.id = itemMusic.songId
                    musicList.add(music)
                }
                playlist.musicList = musicList
                playlists.add(playlist)
            }
            Observable.create(ObservableOnSubscribe<List<Playlist>> {
                it.onNext(playlists)
                it.onComplete()
            })
        }
    }

    /**
     * 获取歌单歌曲
     */
    fun getOnlineSongs(type: String, limit: Int, mOffset: Int): Observable<List<Music>> {
        val params = HashMap<String, String>()

        params[Constants.PARAM_METHOD] = Constants.METHOD_GET_MUSIC_LIST
        params[Constants.PARAM_TYPE] = type
        params[Constants.PARAM_SIZE] = limit.toString()
        params[Constants.PARAM_OFFSET] = mOffset.toString()

        return apiService.getOnlineSongs(params)
                .flatMap { baiduSongList ->
                    val musicList = ArrayList<Music>()
                    for (songInfo in baiduSongList.songList!!) {
                        val music = Music()
                        music.type = Music.Type.BAIDU
                        music.isOnline = true
                        music.id = songInfo.songId
                        music.album = songInfo.albumTitle
                        music.albumId = songInfo.albumId
                        music.artist = songInfo.artistName
                        music.artistId = songInfo.tingUid
                        music.title = songInfo.title
                        music.coverSmall = songInfo.picSmall
                        music.coverUri = songInfo.picBig
                        music.coverBig = songInfo.picRadio
                        musicList.add(music)
                    }
                    Observable.create(ObservableOnSubscribe<List<Music>> { e ->
                        try {
                            e.onNext(musicList)
                            e.onComplete()
                        } catch (ep: Exception) {
                            e.onError(ep)
                        }
                    })
                }
    }

    /**
     * 搜索建议
     */
    fun getSearchSuggestion(query: String): Observable<List<String>> {
        val params = HashMap<String, String>()
        params[Constants.PARAM_METHOD] = Constants.METHOD_SEARCH_SUGGESTION
        params[Constants.PARAM_QUERY] = query
        return apiService.getSearchSuggestion(params)
                .flatMap {
                    Observable.create(ObservableOnSubscribe<List<String>> { e ->
                        try {
                            e.onNext(it.suggestionList!!)
                            e.onComplete()
                        } catch (error: Exception) {
                            e.onError(Throwable(error.message))
                        }
                    })
                }
    }

    /**
     * 获取歌单详情
     * "http://music.baidu.com/data/music/links?songIds=$mid"
     */
    fun getTingSongInfo(mid: String): Observable<Music> {
        val url = Constants.URL_GET_SONG_INFO + mid
        return apiService.getTingSongInfo(url)
                .flatMap { data ->
                    val music = Music()
                    val songInfo = data.data.songList?.get(0)
                    songInfo?.let {
                        music.type = Music.Type.BAIDU
                        music.isOnline = true
                        music.id = songInfo.songId.toString()
                        music.album = songInfo.albumName
                        music.albumId = songInfo.albumId.toString()
                        music.artistId = songInfo.artistId
                        music.artist = songInfo.artistName
                        music.title = songInfo.songName
                        music.uri = songInfo.songLink
                        music.fileSize = songInfo.size.toLong()
                        music.lrcPath = songInfo.lrcLink
                        music.coverSmall = songInfo.songPicSmall
                        music.coverUri = songInfo.songPicBig
                        music.coverBig = songInfo.songPicRadio
                    }
                    Observable.create(ObservableOnSubscribe<Music> { e ->
                        if (music.uri != null) {
                            e.onNext(music)
                            e.onComplete()
                        } else {
                            e.onError(Throwable())
                        }
                    })
                }
    }

    /**
     * 获取歌词
     */
    fun getBaiduLyric(music: Music): Observable<String> {
        //本地歌词路径
        val mLyricPath = FileUtils.getLrcDir() + music.title + "-" + music.artist + ".lrc"
        //网络歌词
        val mLyricUrl = music.lrcPath
        return if (FileUtils.exists(mLyricPath)) {
            Observable.create { emitter ->
                try {
                    val lyric = FileUtils.readFile(mLyricPath)
                    emitter.onNext(lyric)
                    emitter.onComplete()
                } catch (e: Exception) {
                    emitter.onError(e)
                }
            }
        } else apiService.getBaiduLyric(mLyricUrl)
                .flatMap { baiDuLyricInfo ->
                    val lyric = baiDuLyricInfo.string()
                    //保存文件
                    val save = FileUtils.writeText(mLyricPath, lyric)
                    LogUtil.e("保存网络歌词：$save")
                    Observable.fromArray(lyric)
                }
    }

}
